fuel-p2p/src/discovery.rs

```diff
@@ -0,0 +1,498 @@
+use self::mdns::MdnsWrapper;
+use futures::FutureExt;
+use futures_timer::Delay;
+use libp2p::{
+    core::{
+        connection::{ConnectionId, ListenerId},
+        ConnectedPoint,
+    },
+    kad::{handler::KademliaHandlerProto, store::MemoryStore, Kademlia, KademliaEvent, QueryId},
+    mdns::MdnsEvent,
+    swarm::{
+        DialError, IntoProtocolsHandler, NetworkBehaviour, NetworkBehaviourAction, PollParameters,
+        ProtocolsHandler,
+    },
+    Multiaddr, PeerId,
+};
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    io,
+    task::{Context, Poll},
+    time::Duration,
+};
+mod discovery_config;
+mod mdns;
+pub use discovery_config::DiscoveryConfig;
+
+const SIXTY_SECONDS: Duration = Duration::from_secs(60);
+
+/// Event generated by the `DiscoveryBehaviour`.
+#[derive(Debug)]
+pub enum DiscoveryEvent {
+    /// Notify the swarm of a Kademlia event
+    KademliaEvent(KademliaEvent),
+
+    /// Notify the swarm of a connected peer
+    Connected(PeerId),
+
+    /// Notify the swarm of a disconnected peer
+    Disconnected(PeerId),
+}
+
+/// NetworkBehavior for discovery of nodes
+pub struct DiscoveryBehaviour {
+    /// Store the local peer id
+    local_peer_id: PeerId,
+
+    /// Store Peer IDs of currently connected peers
+    connected_peers: HashSet<PeerId>,
```

Comment:
why don't just have `connected_peer_addresses`? less memory, less chance for mistakes (when you update `connected_peer_addresses`, but forgot to update `connected_peers`).

Reply:
good point, tho I updated it a bit, so they now serve a different purpose, `peer_addresses` will hold *any* peers, while the HasheSet will hold IDs of *only* connected peers.

Having said that, it's a work in progress, depending on the use-cases when other behaviours are defined, they might get updated or removed completely.